examples/dft/compile_time_dispatching/complex_fwd_buffer_mklgpu.cpp

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+// STL includes
+#include <iostream>
+
+// oneMKL/SYCL includes
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+#include "oneapi/mkl.hpp"
+
+void run_example(const sycl::device& gpu_device) {
+    constexpr int N = 10;
+
+    // Catch asynchronous exceptions for cpu
+    auto gpu_error_handler = [&](sycl::exception_list exceptions) {
+        for (auto const& e : exceptions) {
+            try {
+                std::rethrow_exception(e);
+            }
+            catch (sycl::exception const& e) {
+                // Handle not dft related exceptions that happened during asynchronous call
+                std::cerr << "Caught asynchronous SYCL exception:" << std::endl;
+                std::cerr << "\t" << e.what() << std::endl;
+            }
+        }
+        std::exit(2);
+    };
+
+    sycl::queue gpu_queue(gpu_device, gpu_error_handler);
+
+    std::vector<std::complex<float>> input_data(N);
+    std::vector<std::complex<float>> output_data(N);
+
+    // enabling
+    // 1. create descriptors
+    oneapi::mkl::dft::descriptor<oneapi::mkl::dft::precision::SINGLE,
+                                 oneapi::mkl::dft::domain::COMPLEX>
+        desc(N);
+
+    // 2. variadic set_value
+    desc.set_value(oneapi::mkl::dft::config_param::PLACEMENT,
+                   oneapi::mkl::dft::config_value::NOT_INPLACE);
+    desc.set_value(oneapi::mkl::dft::config_param::NUMBER_OF_TRANSFORMS,
+                   static_cast<std::int64_t>(1));
+
+    // 3. commit_descriptor (compile_time MKLGPU)
+    desc.commit(oneapi::mkl::backend_selector<oneapi::mkl::backend::mklgpu>{ gpu_queue });
+
+    // 4. compute_forward / compute_backward (MKLGPU)
+    {
+        sycl::buffer<std::complex<float>> input_buffer(input_data.data(), sycl::range<1>(N));
+        sycl::buffer<std::complex<float>> output_buffer(output_data.data(), sycl::range<1>(N));
+        oneapi::mkl::dft::mklgpu::compute_forward<decltype(desc), std::complex<float>,
```

Comment:
The example should not be using internal API, only external API as described in the oneMKL specification.
Why are you using the internal API in the `onapi::mkl::dft::mklgpu` namespace and not the external API `oneapi::mkl::dft::compute_forward`?

Reply:
I (incorrectly) assumed the backend namespaces were intended to be exposed to avoid the overhead in the run-time dispatch mechanism. Looking at the [spec](https://spec.oneapi.io/versions/latest/elements/oneMKL/source/domains/dft/compute_forward.html#onemkl-dft-compute-forward), there doesn't seem to be any mechanism for avoiding the dispatch tables when calling `compute_forward` or `compute_backward` for a known backend. 

I make changes to:
* Change this to `oneapi::mkl::dft::compute_forward`.
* Avoid exposing the internal API.

Reply:
Done in [3715cf2](https://github.com/oneapi-src/oneMKL/pull/259/commits/3715cf2b10bf1c422000e20f4321ebc291898a45).

Reply:
we shouldn't need to include both the shared libraries for one part of the API (compile time dispatch) for the spec to be satisfied and.

Reply:
I see your point and I'm trying to work out what exactly we want here. I've not found guidance in the spec on what should and shouldn't be required from a linking perspective - if there is anything can you point me towards it?

`dft::compute_forward` is part of the `onemkl_dft` library, and `desc.commit(mklgpu_backend)` is part of the `onemkl_dft_mklgpu` library. We need to somehow link against both. There are two options
* Link against both in the CMake explicitly - what we're doing at the moment. As you say, since `onemkl_dft_mklgpu` needs `onemkl` perhaps this should be pulled in automatically.
* Have the `onemkl` dependency pulled in automatically by `onemkl_dft_mklgpu`. This is possible by having `INTERFACE_LINK_LIBRARIES "ONEMKL::SYCL::SYCL;onemkl"` in the cmake of each backend.

However, I'm concerned that you mean that we shouldn't need to link against `onemkl` at all - so neither of the above are satisfactory. Can you clarify what you'd like me to do?

Reply:
So, we would atleast want the `compute_{for|back}ward` symbols in both the libraries, for now they are only in the `onemkl_dft_mklgpu`. AFAIK each library should be independent of its usage. I looked into the rng and blas Cmakes in the examples and I do not see this kind of usage. 

Your implimentation matches the spec, but (maybe later on) it would be good to simplify this kind of usage 
```
oneapi::mkl::dft::compute_forward<decltype(desc), std::complex<float>,
                                                  std::complex<float>>(desc, input_buffer,
                                                                       output_buffer);
```
to 
`oneapi::mkl::dft::compute_forward(desc, input_buffer, output_buffer)`

Reply:
The symbols are now in both libaries in [ef81d21](https://github.com/oneapi-src/oneMKL/pull/259/commits/ef81d212d0faafc32268eea09a6bb4ee6367fbe9).